### ISSUE TITLE
Missing property: `Promise.prototype[Symbol.toStringTag]`

### ISSUE THREAD

**Issue by gibson042** [facebook/hermes#1009]
## Bug Description

https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-promise.prototype-@@tostringtag requires `Promise.prototype` to have a Symbol.toStringTag property with value "Promise", but it is absent in Hermes.

- [x] I have ~~run `gradle clean` and~~ confirmed this bug does not occur with JSC

Hermes version: 0.11.0

## Steps To Reproduce

code example:
```js
Promise.prototype[Symbol.toStringTag]
```

## The Expected Behavior
"Promise"

**Comment by tmikov** [comment]
Hi, thank you for reporting this. 

As described [here](https://github.com/facebook/hermes/blob/e2f48291e4635b230c1267a77c05ea8efe216725/doc/Features.md#miscellaneous-incompatibilities), Hermes implements Promise by using the Promise polyfill https://github.com/then/promise, which unfortunately isn't fully compliant. This was done to ensure compatibility with React Native, which was already using it before Hermes.

Of course this situation isn't ideal, but none of the short term fix options are particularly attractive:
- Develop and maintain a local patch on top of the polyfill.
- Submit a PR to improve the polyfill.
- Find and use a better polyfill.
- Implement a native Promise implementation from first principles.

We also need to confirm that RN does not depend on any non-spec extensions in the existing polyfill. I will reach out to the React Native team to clarify the requirements.

**Comment by ljharb** [comment]
https://npmjs.com/es6-shim is fully compliant, if you wanted to consider an alternative.

**Comment by tmikov** [comment]
@ljharb It is certainly worth looking into. Two quick questions if you don't mind:
- AFAICT, the Promise shim isn't separated. Do you think it would be feasible for us to split it out?
- Do you have an intuition about the performance? My experience with promise is from 9 years ago, and at that time, IIRC, polyfills had wildly different performance. I think Bluebird was the fastest at the time.

**Comment by ljharb** [comment]
You're right that it's not separated; I would be happy to split it out if that helps the hermes team.

Bluebird remains the fastest by violating the spec in a number of small ways; I'm not sure about the performance at this point tbh, but I'm also open to performance improvements as long as they don't violate the spec.

**Comment by tmikov** [comment]
@ljharb offer much appreciated, thank you! I will reach out in the es6-shim repo if we need help. Changing the Promise implementation is a very risky process, I don't think we have any benchmarks for it in place yet. Previously we simply took the polyfill already in use, so there was no need to worry about perf or functionality regressions.

**Comment by gibson042** [comment]
I don't know if anything has changed since the issue was opened, but it looks like this should be relatively easy to fix by adding something to [lib/InternalBytecode/01-Promise.js](https://github.com/facebook/hermes/blob/8dc886dabfe2fa90614294b23d08c845243300b8/lib/InternalBytecode/01-Promise.js) like
```js
Object.defineProperty(core.prototype, Symbol.toStringTag, {
  value: 'Promise',
  writable: false,
  enumerable: false,
  configurable: true,
});
```

**Comment by tmikov** [comment]
As was mentioned before, unfortunately we can't easily patch 01-Promise.js, since it is automatically generated from https://www.npmjs.com/package/promise by https://github.com/facebook/hermes/blob/8dc886dabfe2fa90614294b23d08c845243300b8/utils/gen-promise-internal-bc.sh .

We were forced to use it for compatibility with React Native, otherwise we would rather use https://www.npmjs.com/package/es6-shim.

@gibson042 are you saying that React Native uses a different polyfill or no polyfill at all when running with JSC? If that is indeed the case, then perhaps the compatibility isn't so critical and we can switch to a better version.

**Comment by grifotv** [comment]
@tmikov what would be the ideal way to apply a Promise polyfill in a React Native app with Hermes enabled?

For instance, I tried using https://github.com/ungap/with-resolvers and here are the results I got:
- Debug mode with Hermes enabled = works fine
- Debug mode with Hermes disabled = works fine
- Release mode with Hermes enabled = **doesn't work**
- Release mode with Hermes disabled = works fine

**Comment by tmikov** [comment]
@grifotv if you believe there is a problem in Hermes, please create a minimal repro without React Native, submit a new issue (since this issue specifically is about `Promise.prototype[Symbol.toStringTag]`, and we will take a look.

We (the Hermes team) can't provide meaningful help about React Native, we are only focusing on the JS engine.

**Comment by tmikov** [comment]
@ljharb we discussed it at length and decided it is high time we became spec-compliant as much as possible. Does your offer to split the Promise shim still stand? We would love to use it.

**Comment by ljharb** [comment]
Absolutely! I’ll see if i can get that done in the next week or so.

**Comment by leotm** [comment]
> We also need to confirm that RN does not depend on any non-spec extensions in the existing polyfill. I will reach out to the React Native team to clarify the requirements.

> @gibson042 are you saying that React Native uses a different polyfill or no polyfill at all when running with JSC? If that is indeed the case, then perhaps the compatibility isn't so critical and we can switch to a better version.

React Native (JSC) uses [then/promise](https://github.com/then/promise)
- `promise/setimmediate/es6-extensions`
- `promise/setimmediate/finally`
- `promise/setimmediate/rejection-tracking`

```js
// github.com/facebook/react-native/blob/main/packages/react-native/Libraries/Core/polyfillPromise.js

// ...

const {polyfillGlobal} = require('../Utilities/PolyfillFunctions');

/**
 * Set up Promise. The native Promise implementation throws the following error:
 * ERROR: Event loop not supported.
 *
 * If you don't need these polyfills, don't use InitializeCore; just directly
 * require the modules you need from InitializeCore for setup.
 */

// If global.Promise is provided by Hermes, we are confident that it can provide
// all the methods needed by React Native, so we can directly use it.
if (global?.HermesInternal?.hasPromise?.()) {
  const HermesPromise = global.Promise;

  if (__DEV__) {
    if (typeof HermesPromise !== 'function') {
      console.error('HermesPromise does not exist');
    }
    global.HermesInternal?.enablePromiseRejectionTracker?.(
      require('../promiseRejectionTrackingOptions').default,
    );
  }
} else {
  polyfillGlobal('Promise', () => require('../Promise'));
}
```

```js
// github.com/facebook/react-native/blob/main/packages/react-native/Libraries/Promise.js

// ...

const Promise = require('promise/setimmediate/es6-extensions');

require('promise/setimmediate/finally');

if (__DEV__) {
  require('promise/setimmediate/rejection-tracking').enable(
    require('./promiseRejectionTrackingOptions').default,
  );
}

module.exports = Promise;
```

same as Hermes

https://github.com/facebook/hermes/blob/568b1c99be3560e9634d04c1213813049dcfce6f/utils/promise/index.js#L1-L15

so yes unfortunately isn't fully compliant

**Comment by ljharb** [comment]
(I'm still working on this fwiw, but it's turning out to be quite complex to polyfill Promise in a non-automatically-installed kind of way)

**Comment by leotm** [comment]
> @tmikov what would be the ideal way to apply a Promise polyfill in a React Native app with Hermes enabled?

```js
// github.com/facebook/react-native/blob/main/packages/react-native/Libraries/Core/polyfillPromise.js

// ...

// Apply some other Promise polyfill only on Hermes
if (global?.HermesInternal) {
  require('@ungap/with-resolvers');
}

// ...
```

**Comment by lavenzg** [comment]
This has been added in static_h branch.
